README.md
# Bouncing ball
//This is a simple game where you have a timer that ticks down, and you can click the bouncing ball to add time to your timer and score points.

//How to Play
//Use the mouse to click on the bouncing ball whenever it appears on the screen.
//Each time you click the ball, your timer will be increased, and your score will go up.
//The game ends when the timer runs out.
//Technologies Used
//This game was created using HTML5 Canvas and pure JavaScript for the game logic and rendering.

//Acknowledgments
//I would like to acknowledge the following resources that helped me in creating this game:

//MDN Web Docs - 2D Breakout Game Pure JavaScript Tutorial - This guide from MDN Web Docs helped me learn how to use the HTML5 Canvas to draw and create the walls and bouncing ball in the game.

//YouTube - Click Event Listener Tutorial - This video tutorial helped me understand how to use the click event listener, which allowed me to implement the ball-clicking mechanism in the game.

//YouTube - Understanding Canvas and Drawing - This video helped me gain a better understanding of how to work with HTML5 Canvas for drawing and interacting with objects on the canvas.

//Image Credits
//I used images from Peakpx for the game. Unfortunately, I couldn't find the specific artist for the images I used.

//How to Contribute
//If you would like to contribute to this project, feel free to submit bug reports, feature requests, or even code contributions. Please create a pull request, and I'll be happy to review it.

//License
//This project is licensed under the MIT License.

//This revised README provides clearer instructions on how to play the game, highlights the technologies and resources used, and properly credits the sources of information and images used. It also includes sections for contributions and licensing, which are essential in open-source projects. Remember that a good README should be concise, informative, and well-organized, making it easier for others to understand and contribute to your project.


CHANGELOG.md
//[Vison-0.7 7/27/23]

//Added
//-Created a box environment to contain the ball movement.
//-Implemented the ball movement within the box.
//Improved
//-Calculated the center of the ball for smooth movement using a math equation.
//Fixed
//-Fixed an issue where the ball wouldn't register clicks. Calculated the ball's movement within the box and assigned it a center, resolving the problem.
//Removed
//-[If you removed any features or code, mention them here]


draw.ts

function initGame(): void {
const canvas = document.getElementById("canvas1") as HTMLCanvasElement;
const context = canvas.getContext("2d");
if(!context){
  return console.log("failted to get 2d from html");;
}else { 
const scoreElement = document.getElementById('score') as HTMLElement;
const tim = document.getElementById('timer') as HTMLElement;

const CANVAS_WIDTH = canvas.width = 150;
const CANVAS_HEIGHT = canvas.height = 150;

let ballSize: number = 25;
let placeSide: number = canvas.width / 2;
let placeTop: number = canvas.height - 30;
let moveSide: number = 2;
let moveTop: number = -2;

let currentScore: number = 0;
let timeLeft: number = 10; // Number of seconds before the game ends

function updateScore<T>(score: T): void {
  scoreElement.innerText = `Score: ${score} (Clicks: ${clickCount})`;
}

let clickCount: number = 0;

function clickCounter(x: number, y: number): boolean {
  const distance = Math.sqrt((x - placeSide) ** 2 + (y - placeTop) ** 2);
  if (distance < ballSize) {
    clickCount++;
    currentScore += 1;
    timeLeft += 5;
    updateScore(currentScore); // Using generic function to update the score
    scoreElement.innerText = `Score: ${currentScore} (Clicks: ${clickCount})`;
    return true;
  } else {
    return false;
  }
}

canvas.addEventListener('click', (event: MouseEvent) => {
  const rect = canvas.getBoundingClientRect();
  const x = event.clientX - rect.left;
  const y = event.clientY - rect.top;
  console.log(clickCounter(x, y));
});

function drawingBall(): void {
  context.beginPath();
  context.arc(placeSide, placeTop, ballSize, 0, Math.PI * 2);
  context.fillStyle = "red";
  context.fill();
  context.closePath();
}

function moveBall(): void {
  context.clearRect(5, 5, canvas.width, canvas.height);
  drawingBall();
  if (placeSide + moveSide > canvas.width - ballSize || placeSide + moveSide < ballSize) {
    moveSide = -moveSide;
  }
  if (placeTop + moveTop > canvas.height - ballSize || placeTop + moveTop < ballSize) {
    moveTop = -moveTop;
  }
  placeSide += moveSide;
  placeTop += moveTop;
}

function updateTimer(): void {
  timeLeft--;
  tim.innerText = `Time left: ${timeLeft} seconds`;

  if (timeLeft === 0) {
    clearInterval(timer);
    canvas.removeEventListener('click', clickHandler);
    clearInterval(ballInterval);
    context.clearRect(0, 0, canvas.width, canvas.height);
    context.font = "30px Arial";
    context.textAlign = "center";
    context.fillText(`Game over! Your score is ${currentScore} (Clicks: ${clickCount})`, canvas.width / 2, canvas.height / 2);
  }
}

function clickHandler(event: MouseEvent): void {
  const rect = canvas.getBoundingClientRect();
  const x = event.clientX - rect.left;
  const y = event.clientY - rect.top;
  clickCounter(x, y);
}

canvas.addEventListener('click', clickHandler);

const ballInterval = setInterval(moveBall, 19);

// Start the timer
const timer = setInterval(updateTimer, 1000);
updateScore(currentScore); // Update the score once at the beginning using generic function
}
}
initGame()

draw.js
function initGame() {
    var canvas = document.getElementById("canvas1");
    var context = canvas.getContext("2d");
    if (!context) {
        return console.log("failted to get 2d from html");
        ;
    }
    else {
        var scoreElement_1 = document.getElementById('score');
        var tim_1 = document.getElementById('timer');
        var CANVAS_WIDTH = canvas.width = 150;
        var CANVAS_HEIGHT = canvas.height = 150;
        var ballSize_1 = 25;
        var placeSide_1 = canvas.width / 2;
        var placeTop_1 = canvas.height - 30;
        var moveSide_1 = 2;
        var moveTop_1 = -2;
        var currentScore_1 = 0;
        var timeLeft_1 = 10; // Number of seconds before the game ends
        function updateScore(score) {
            scoreElement_1.innerText = "Score: ".concat(score, " (Clicks: ").concat(clickCount_1, ")");
        }
        var clickCount_1 = 0;
        function clickCounter(x, y) {
            var distance = Math.sqrt(Math.pow((x - placeSide_1), 2) + Math.pow((y - placeTop_1), 2));
            if (distance < ballSize_1) {
                clickCount_1++;
                currentScore_1 += 1;
                timeLeft_1 += 5;
                updateScore(currentScore_1); // Using generic function to update the score
                scoreElement_1.innerText = "Score: ".concat(currentScore_1, " (Clicks: ").concat(clickCount_1, ")");
                return true;
            }
            else {
                return false;
            }
        }
        canvas.addEventListener('click', function (event) {
            var rect = canvas.getBoundingClientRect();
            var x = event.clientX - rect.left;
            var y = event.clientY - rect.top;
            console.log(clickCounter(x, y));
        });
        function drawingBall() {
            context.beginPath();
            context.arc(placeSide_1, placeTop_1, ballSize_1, 0, Math.PI * 2);
            context.fillStyle = "red";
            context.fill();
            context.closePath();
        }
        function moveBall() {
            context.clearRect(5, 5, canvas.width, canvas.height);
            drawingBall();
            if (placeSide_1 + moveSide_1 > canvas.width - ballSize_1 || placeSide_1 + moveSide_1 < ballSize_1) {
                moveSide_1 = -moveSide_1;
            }
            if (placeTop_1 + moveTop_1 > canvas.height - ballSize_1 || placeTop_1 + moveTop_1 < ballSize_1) {
                moveTop_1 = -moveTop_1;
            }
            placeSide_1 += moveSide_1;
            placeTop_1 += moveTop_1;
        }
        function updateTimer() {
            timeLeft_1--;
            tim_1.innerText = "Time left: ".concat(timeLeft_1, " seconds");
            if (timeLeft_1 === 0) {
                clearInterval(timer_1);
                canvas.removeEventListener('click', clickHandler);
                clearInterval(ballInterval_1);
                context.clearRect(0, 0, canvas.width, canvas.height);
                context.font = "30px Arial";
                context.textAlign = "center";
                context.fillText("Game over! Your score is ".concat(currentScore_1, " (Clicks: ").concat(clickCount_1, ")"), canvas.width / 2, canvas.height / 2);
            }
        }
        function clickHandler(event) {
            var rect = canvas.getBoundingClientRect();
            var x = event.clientX - rect.left;
            var y = event.clientY - rect.top;
            clickCounter(x, y);
        }
        canvas.addEventListener('click', clickHandler);
        var ballInterval_1 = setInterval(moveBall, 19);
        // Start the timer
        var timer_1 = setInterval(updateTimer, 1000);
        updateScore(currentScore_1); // Update the score once at the beginning using generic function
    }
}
initGame();
